### ISSUE TITLE
Onitu on local storage on windows is blocking after uploading file 

### ISSUE THREAD

**Issue by iloria-iloria** [onitu/onitu#74]
Hello,
I've ported the move, delete, renaming on windows, it is working without any problem when the file is uploaded at the root of the folder, but each time i upload/rename/move a file in a subfolder, the operation is successful but Onitu then stay blocked somewhere and i can't find where.

According to the research i did, there may be an issue with doing a return in the windows loop line 256 of `local_storage.py`.
I tried using a continue or a break instead, it seems to fix part of the problem as the drivers seems to manage to notify the referee about a new file, contrary to using a return.But the file is still not uploaded to the other drivers and i can't figure why.

**Comment by k4nar** [comment]
Yeah, it should be a `continue`, I'll fix that. Expect this bug I don't see what could happen here. Could you show me some debug ?

**Comment by iloria-iloria** [comment]
Hello,
Sorry for the delay, i din't have proper internet access until now.
I had lot of debug that showed me that it was the receiving driver driver that was blocking. 
when cleaning my debug to make it comprehensive, i noticed that drivers B is notifying several time the referee for the same file. searching further, it seems that i'm going out of win32file.ReadDirectoryChangesW  several time in a row for each file added, 

my current Debug where test is displayed just before entering win32file.ReadDirectoryChangesW: 

```
[2014-08-11 21:54] INFO: Onitu: Loading setup...
[2014-08-11 21:54] INFO: Escalator: Starting on 'tcp://127.0.0.1:21969'
[2014-08-11 21:54] DEBUG: Onitu: Loading entry B
[2014-08-11 21:54] DEBUG: Onitu: Loading entry A
[2014-08-11 21:54] DEBUG: Onitu: Entries loaded
[2014-08-11 21:54] INFO: Referee: Started
[2014-08-11 21:54] INFO: A: Started
[2014-08-11 21:54] INFO: B: Started
[2014-08-11 21:54] DEBUG: B: test
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] INFO: B - Router: Started
[2014-08-11 21:54] INFO: B - Dealer: Started
[2014-08-11 21:54] INFO: A - Router: Started
[2014-08-11 21:54] INFO: A - Dealer: Started
[2014-08-11 21:54] INFO: REST API: Starting on localhost:3862
[2014-08-11 21:54] DEBUG: B: Notifying the Referee about 'New folder\file.txt'
[2014-08-11 21:54] DEBUG: B: test
[2014-08-11 21:54] DEBUG: B: Notifying the Referee about 'New folder\file.txt'
[2014-08-11 21:54] INFO: Referee: Update for '/New folder\file.txt' from B
[2014-08-11 21:54] DEBUG: B: test
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] INFO: A - Dealer: Starting to get 'New folder\file.txt' from B
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] DEBUG: A - Dealer: Waiting for ROUTER port for B
[2014-08-11 21:54] DEBUG: B - Router: Getting chunk of size 1048576 from offset 0 in 'New folder\file.txt'

[2014-08-11 21:54] DEBUG: A - Dealer: Got ROUTER port for B
[2014-08-11 21:54] DEBUG: A - Dealer: Connected
[2014-08-11 21:54] DEBUG: A - Dealer: Asking B for a new chunk
[2014-08-11 21:54] DEBUG: A - Dealer: Received chunk of size 13 from B for 'New folder\file.txt'
[2014-08-11 21:54] INFO: Referee: Update for '/New folder\file.txt' from A
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:54] INFO: A - Dealer: Transfer of 'New folder\file.txt' from B successful
[2014-08-11 21:54] DEBUG: A: Notifying the Referee about 'New folder\file.txt'
[2014-08-11 21:54] DEBUG: A: test
[2014-08-11 21:56] DEBUG: B: Notifying the Referee about 'New folder\file 2.txt'
[2014-08-11 21:56] DEBUG: B: test
[2014-08-11 21:56] DEBUG: B: Notifying the Referee about 'New folder\file 2.txt'
[2014-08-11 21:56] DEBUG: B: test
```

I've tried searching for alternative to our current method to detect file change juste in case it really come from there and found a library called watchDog which seems to be a cross-plateform library to detect file change, is there any reason we aren't using it, at least on windows ?

**Comment by k4nar** [comment]
Well, maybe it would help to print everything you get from `win32file.ReadDirectoryChangesW` in order to understand why you are detecting the same file several times. That said, I don't see why it could be such an issue, as the Referee should be able to handle it.

Regarding Watchdog, we were using it (we even contributed to it), but there was several issues on Linux and Mac OS which made it useless. Maybe you could try it again for Windows, the API is pretty simple.

**Comment by iloria-iloria** [comment]
it seems that my current issue come from issue #77 since file detection on windows is way messier with our current method, driver A notify the referee for a new file after receiving a file from B, which trigger #77.

**Comment by k4nar** [comment]
Can you try again ?

**Comment by iloria-iloria** [comment]
it seems that it was part of the issue again. if i work on my branch and cherry pick on the fix for #77, i get back to the point i was at the end of july when i opened this issue. if i cherry pick the fix for #77 + the fix for this issue (replacing return by continue), i get another problem : 

```
[2014-08-21 09:31] INFO: Onitu: Loading setup...
[2014-08-21 09:31] INFO: Escalator: Starting on 'tcp://127.0.0.1:26846'
[2014-08-21 09:31] DEBUG: Onitu: Loading entry A
[2014-08-21 09:31] DEBUG: Onitu: Loading entry B
[2014-08-21 09:31] DEBUG: Onitu: Entries loaded
[2014-08-21 09:31] INFO: Referee: Started
[2014-08-21 09:31] INFO: A: Started
[2014-08-21 09:31] INFO: A - Router: Started
[2014-08-21 09:31] INFO: A - Dealer: Started
[2014-08-21 09:31] INFO: REST API: Starting on localhost:3862
[2014-08-21 09:31] INFO: B: Started
[2014-08-21 09:31] INFO: B - Router: Started
[2014-08-21 09:31] INFO: B - Dealer: Started
[2014-08-21 09:32] DEBUG: B: update
[2014-08-21 09:32] DEBUG: B: Notifying the Referee about 'card cancel numver.txt'
[2014-08-21 09:32] INFO: Referee: Update for '/card cancel numver.txt' from B
[2014-08-21 09:32] DEBUG: B: update
[2014-08-21 09:32] DEBUG: B: Notifying the Referee about 'card cancel numver.txt'
[2014-08-21 09:32] DEBUG: B: update
[2014-08-21 09:32] DEBUG: B: Notifying the Referee about 'card cancel numver.txt'
[2014-08-21 09:32] INFO: A - Dealer: Starting to get 'card cancel numver.txt' from B
[2014-08-21 09:32] DEBUG: A - Dealer: Waiting for ROUTER port for B
[2014-08-21 09:32] DEBUG: A - Dealer: Got ROUTER port for B
[2014-08-21 09:32] DEBUG: A - Dealer: Connected
[2014-08-21 09:32] DEBUG: A - Dealer: Asking B for a new chunk
[2014-08-21 09:32] DEBUG: B - Router: Getting chunk of size 1048576 from offset 0 in 'card cancel numver.txt'
[2014-08-21 09:32] DEBUG: A - Dealer: Received chunk of size 13 from B for 'card cancel numver.txt'
[2014-08-21 09:32] INFO: A - Dealer: Transfer of 'card cancel numver.txt' from B successful
[2014-08-21 09:32] DEBUG: A: update
[2014-08-21 09:32] DEBUG: A: Notifying the Referee about 'card cancel numver.txt'
[2014-08-21 09:32] INFO: Referee: Update for '/card cancel numver.txt' from A
[2014-08-21 09:32] DEBUG: Referee: The file '/card cancel numver.txt' on B is no longer under onitu control. should be d
eleted.
```

strangly, right after upload, the file on B is no longer under the control of onitu, i'm wondering if it is because of A notifying the referee and him then saying eh, that's not on the good side, let's ignore this. I'm not sure of how it works here, can you check this particular point ? i'll continue investigating meanwhile.

**Comment by k4nar** [comment]
I'm seeing a few bugs with the `owners` field. The message from the Referee might be related to that. Anyway, it shouldn't be an issue for you.

**Comment by iloria-iloria** [comment]
are you sure it can't be an issue for me ?

 most of the move/delete function do this : 

```
if plug.name not in metadata.owners:
    return
```

and it return everytime there, since the no longer under onitu control is also verifying something related to owner i though it was related.

edit:

i think i may have an idea of what is going on here : 

as i explained before, file detection currently is quite bad : each file created is notified several time, even when it's a file that was received through onitu. what happen is when a new file is detected on B, it is uploaded on A and then A notify the referee about a new file.
This should not be a problem in the case it is a folder that is also synced by onitu. but it become one when the folder isn't.

my guess is that when referee is running its usual check, it run this sets of code : 

```
        for rule in self.rules:
            if self.rule_match(rule, filename):
                should_own.update(rule.get("sync", []))
                should_own.difference_update(rule.get("ban", []))
```

and when it gets to the last line, since the file is not in a synced folder, it remove it from the owner.

**Comment by k4nar** [comment]
I think we should focus on the main issue first: On Windows, notifications are triggered several times.
Once that's solved, we'll start to investigate what's wrong with the Referee, but it's being rewritten anyway :) .

**Comment by iloria-iloria** [comment]
in order to investigate why several notifcation are triggered each time i isolated the code responsible of the file detection out of onitu (i should have done that lot earlier in fact). since i didn't have the cross debug from 2 drivers i got a better view of what was going on : 
- multiple file notification doesn't trigger when a file is creating, in fight if a single empty file is created, i just get one notification (i was always copy pasting files, so i didn't really notice before)
- however if i copy paste a file, i get a creation notification and then i get multiple modification notification the bigger the file the more notification i usually get.

so i investigated what was going on and it seems the win32con.FILE_NOTIFY_CHANGE_ATTRIBUTES is responsible for all the notification. which mean that when copying a new file, for some reason one of the flag is constantly changing. do we have any important things going on with the attribute (maybe read only or hidden attribute ?) or can i just remove this flag ?

so there is one last problem : when i copy a file i get two notif, one for the creation of the file and another when the actual content of the file is copied. from there i guess we have two approach : 
- notify referee anyway since copying an empty file should not be a problem, and then upload the real file on the second notif
- try to ignore the file creation notification. once again i think think there is two approach for that : 
  - ignore empty file, even if it is a legit empty file
  - find a way to determine if the file is being copied/modified by an external file. i'm not exactely sure how to do that properly yet, but i'm looking into it.

I also noticed while running test that when replacing file through windows or other prog i get two nofication on the same model than a file copy.

**Comment by k4nar** [comment]
I don't remember why we are using `FILE_NOTIFY_CHANGE_ATTRIBUTES`. Maybe it's for the case when an hidden file becomes visible (we may not want to address this issue), maybe it's for another reason but I don't see any.

For your second issue, maybe a solution would be to collapse two events for the same file very close in time (like less than 1s). It requires a bit of coding, but it's not that hard and seems to be a pretty good solution (only delays the notifications by 1s).

**Comment by iloria-iloria** [comment]
I finally was able to get a working version of the local storage on windows (30e2bf3), i had to manually check quite a few things, it needs more testing and i need to clean the code, but i've been able to work around the issue i got : 
- i'm collapsing all write notification for a file until i don't get notification for this file for 1 sec. this is necessary for small file where os.access(abs_path, os.R_OK) checks occurs after the closing of the file.
- to b able to collapse notification, I went from treating the notification on a synchronous manner to an asynchronous manner. This is also a great preparation for the new rules since this way will permit to handle watching multiple folder a much more easily.
- i'm able to ignore early notification for big files thanks to os.access(abs_path, os.R_OK) however, i get true to this condition too early and end up having to verify if i can open the file before sending the notification : 

```
try:
     fd = os.open(i, os.O_RDONLY)
except: 
     continue   <-  i have a  for earlier, since i could not open, the file is not ready yet 
else:
    [..] # send notification ect
```

Before you ask, yes i've tried to stat it and i was able to do it while the open failed, but if you know other method to check that i could have forgot tell me.
- ignoring extra notification when moving onitu-tmp file and when receiving files is done through 2 dict currentry. This is the part that most need to be cleaned currently and i will review it shortly, but the behaviour should still be the same.
